clusterstate/provider/aws.go

```diff
@@ -0,0 +1,227 @@
+	{
+		c := apprclient.Config{
+			Fs:     afero.NewOsFs(),
+			Logger: a.logger,
+
+			Address:      CNRAddress,
+			Organization: CNROrganization,
+		}
+
+		apprClient, err = apprclient.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var helmClient *helmclient.Client
+	{
+		c := helmclient.Config{
+			Logger:    a.logger,
+			K8sClient: a.guestFramework.K8sClient(),
+
+			RestConfig: a.guestFramework.RestConfig(),
+		}
+
+		helmClient, err = helmclient.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = helmClient.EnsureTillerInstalled()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	// Install the e2e app chart in the guest cluster.
+	{
+		a.logger.Log("level", "debug", "message", "installing e2e-app for testing")
+
+		tarballPath, err := apprClient.PullChartTarball(ChartName, ChartChannel)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = helmClient.InstallFromTarball(tarballPath, ChartNamespace)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (a *AWS) RebootMaster() error {
+	masterInstanceName := fmt.Sprintf("%s-master", a.clusterID)
+	describeInput := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("tag:Name"),
+				Values: []*string{
+					aws.String(masterInstanceName),
+				},
+			},
+			{
+				Name: aws.String("instance-state-name"),
+				Values: []*string{
+					aws.String("running"),
+				},
+			},
+		},
+	}
+	res, err := a.awsClient.EC2.DescribeInstances(describeInput)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	if len(res.Reservations) == 0 {
+		return microerror.Maskf(notFoundError, "ec2 instance %q not found", masterInstanceName)
+	}
+	if len(res.Reservations) > 1 {
+		return microerror.Maskf(tooManyResultsError, "found %d running instances named %q", len(res.Reservations), masterInstanceName)
+	}
+
+	masterID := res.Reservations[0].Instances[0].InstanceId
+	rebootInput := &ec2.RebootInstancesInput{
+		InstanceIds: []*string{
+			masterID,
+		},
+	}
+	_, err = a.awsClient.EC2.RebootInstances(rebootInput)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (a *AWS) ReplaceMaster() error {
+	customObject, err := a.hostFramework.G8sClient().ProviderV1alpha1().AWSConfigs("default").Get(a.clusterID, metav1.GetOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Change instance type to trigger replacement of existing master node.
+	customObject.Spec.AWS.Masters[0].InstanceType = EC2InstanceType
+
+	_, err = a.hostFramework.G8sClient().ProviderV1alpha1().AWSConfigs("default").Update(customObject)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (a *AWS) WaitForAPIDown() error {
+	err := a.guestFramework.WaitForAPIDown()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (a *AWS) WaitForGuestReady() error {
+	err := a.guestFramework.WaitForGuestReady()
```

Comment:
I also think this should be independent of the provider. Together with installing the app.